seq2science/util.py

```diff
@@ -407,3 +408,132 @@ def shorten(string, max_length, methods="right"):
+    if 'descriptive_name' in samples:
+        samples['descriptive_name'] = samples['descriptive_name'].mask(pd.isnull, samples['replicate']) if \
+            'replicate' in samples else samples['descriptive_name'].mask(pd.isnull, samples['sample'])
+        if ('replicate' in samples and samples['descriptive_name'].to_list() == samples['replicate'].to_list()) or \
+                samples['descriptive_name'].to_list() == samples['sample'].to_list():
+            samples = samples.drop(columns=['descriptive_name'])
+    if 'strandedness' in samples:
+        samples['strandedness'] = samples['strandedness'].mask(pd.isnull, 'nan')
+        if config.get('ignore_strandedness', True) or not any(
+                [field in list(samples['strandedness']) for field in ['yes', 'forward', 'reverse', 'no']]):
+            samples = samples.drop(columns=['strandedness'])
+    if 'colors' in samples:
+        samples['colors'] = samples['colors'].mask(pd.isnull, '0,0,0')  # nan -> black
+        samples['colors'] = [color_parser(c) for c in samples['colors']]  # convert input to HSV color
+        if not config.get('create_trackhub', False):
+            samples = samples.drop(columns=['colors'])
+
+    if 'replicate' in samples:
+        # check if replicate names are unique between assemblies
+        r = samples[['assembly', 'replicate']].drop_duplicates().set_index('replicate')
+        for replicate in r.index:
+            assert len(r[r.index == replicate]) == 1, \
+                ("\nReplicate names must be different between assemblies.\n" +
+                 f"Replicate name '{replicate}' was found in assemblies {r[r.index == replicate]['assembly'].tolist()}.")
+
+    # check if sample, replicate and condition names are unique between the columns
+    for idx in samples.index:
+        if "condition" in samples:
+            assert idx not in samples["condition"].values, f"sample names, conditions, and replicates can not overlap. " \
+                                                           f"Sample {idx} can not also occur as a condition"
+        if "replicate" in samples:
+            assert idx not in samples["replicate"].values, f"sample names, conditions, and replicates can not overlap. " \
+                                                           f"Sample {idx} can not also occur as a replicate"
+
+    if "condition" in samples and "replicate" in samples:
+        for cond in samples["condition"]:
+            assert cond not in samples[
+                "replicate"].values, f"sample names, conditions, and replicates can not overlap. " \
+                                     f"Condition {cond} can not also occur as a replicate"
+
+    # validate samples file
+    for schema in sample_schemas:
+        validate(samples, schema=f"{config['rule_dir']}/../schemas/samples/{schema}.schema.yaml")
+
+    sanitized_samples = copy.copy(samples)
+
+    samples = samples.set_index('sample')
+    samples.index = samples.index.map(str)
```

Comment:
if this becomes an external function, shouldn't `samples` be returned, and `sanitized_samples` be copied outside this function?

On a different note: do we need a separate `sanitized_samples`?

Reply:
Not sure if we need a separate one.. It's not really used